Add structure and API

msricher opened the pull request:
@matt-chan can this be merged?  We need some sort of structure and API to continue.

msricher committed:
Add skeleton/structure for additional features

* Minimize routine
* Additional quasi-Newton methods
* Gradient & Hessian classes
* Line Search
* Trust Region

make setup.py work

msricher committed:
Add Bofill & PSB quasi-Newton update

msricher committed:
fixup routines + line search and tr API

msricher commented:
:( coverage fails.... It's going to do that.

msricher commented on flik/__init__.py:
```diff
@@ -25,17 +25,70 @@
+
+# __version__ definition should be the last line of this file,
+# because the setup.py script uses it to determine the version
+__version__ = "0.0.0"
```
This allows users to actually install via setup.py.

msricher commented on flik/linesearch/base.py:
```diff
@@ -0,0 +1,79 @@
+    ]
+
+
+class LineSearch:
```
Specific line search methods would subclass LineSearch

msricher commented on flik/methods/base.py:
```diff
@@ -0,0 +1,789 @@
+    ]
+
+
+class Method:
```
Basic quasi-Newton methods go here.  Anything requiring keyword options gets a subclass.

msricher commented on flik/methods/bofill.py:
```diff
@@ -0,0 +1,93 @@
+    ]
+
+
+class BofillMethod(Method):
```
Here's an example of a parameterized method.

msricher commented on flik/trustregion/base.py:
```diff
@@ -0,0 +1,73 @@
+    ]
+
+
+class TrustRegion:
```
Specific trust region methods would subclass TrustRegion

msricher commented on setup.py:
```diff
@@ -29,31 +30,33 @@
 
-def get_version():
-    """Load the version from version.py, without importing it.
+def version():
```
Made this read version string from final line of \_\_init\_\_.py so that users can install via setup.py.

msricher replied:
"users" being my scrum group :)

msricher commented on setup.py:
```diff
@@ -64,12 +67,12 @@ def readme():
         author='Ayers Lab',
         author_email='ayers@mcmaster.ca',
         url='https://github.com/QuantumElephant/Flik',
-        packages=['flik', 'flik.test'],
```
?? Just cleaned this up, 'flik' will get all submodules

msricher commented on setup.py:
```diff
@@ -29,31 +30,33 @@
 
 import sys
+
+from io import open
```
`from io import open` gets Python3-enabled `open` on Python2, and does nothing on Python3.  At least allows users running setup.py from Python2 to get into the warning and setup() call.

msricher commented on flik/optimize/functions.py:
```diff
@@ -0,0 +1,512 @@
+    return _root_iteration(f, x, J, tol, maxiter, method, linesearch, trustregion)
+
+
+def minimize(f, x, grad=None, H=None, tol=1.0e-9, maxiter=100, method=None,
```
It's the same code as `root`...

msricher commented on flik/optimize/functions.py:
```diff
@@ -0,0 +1,512 @@
+    trustregion = _check_trustregion(trustregion, grad, H, tropts)
+    # Minimization problem is equivalent to root-finding with
+    # cost = minimization target t, f == grad(t), and J == H(t).
+    result = _root_iteration(grad, x, H, tol, maxiter, method, linesearch, trustregion)
```
...exact same routine

msricher commented on flik/gradient/__init__.py:
```diff
@@ -0,0 +1,95 @@
+    ]
+
+
+class Gradient(Objective):
```
All gradients can be objective functions for root finding.  Not all objective functions for root finding are gradients.

msricher commented on flik/hessian/__init__.py:
```diff
@@ -0,0 +1,79 @@
+    ]
+
+
+class Hessian(Jacobian):
```
All Hessians are the Jacobian of their corresponding gradient function.  Not all Jacobians are Hessians.

msricher commented on flik/jacobian.py:
```diff
@@ -1,214 +0,0 @@
-        """
-        return self._jac(x)
-
-    def update_newton(self, A, new_x, *_):
```
Paul wants like 100 different quasi-newton updates, so I moved these to flik/methods/

msricher commented on flik/optimize/exceptions.py:
```diff
@@ -0,0 +1,40 @@
+    ]
+
+
+class NoConvergenceError(LinAlgError):
```
This subclass will save us trouble later (every other exception for line search, trust region, etc. should subclass this).

msricher commented on flik/jacobian/base.py:
```diff
@@ -0,0 +1,66 @@
+    ]
+
+
+class Jacobian:
```
Ideas -- add more FiniteDiffJacobian subclasses.  3-point finite-difference, for example.

msricher commented on flik/jacobian/__init__.py:
```diff
@@ -0,0 +1,32 @@
+
+
+from flik.jacobian.base import Jacobian
+from flik.jacobian.finitediff import ForwardDiffJacobian
```
Ideas -- add more FiniteDiff{Jacobian,Hessian} subclasses. 3-point finite-difference, for example.
              Also, we could add in FiniteDiff{Objective,Gradient} subclasses like 3-point.  
              Better approximations  than finite-difference would also be nice.

msricher replied:
Automatic code differentiation, or symbolic code integration via SymPy, would be cool too.  AutoGradient, AutoHessian, SymbolicGradient, SymbolicHessian.

msricher replied:
https://github.com/HIPS/autograd
I'm not sure if you or Paul sent me this.

msricher commented on flik/linesearch/backtrace.py:
```diff
@@ -0,0 +1,100 @@
+    ]
+
+
+class BacktraceLineSearch(ConstantLineSearch):
```
Quick example of backtracing line search implementation

msricher commented on flik/linesearch/constant.py:
```diff
@@ -0,0 +1,83 @@
+    ]
+
+
+class ConstantLineSearch(LineSearch):
```
Line search that just scales step size by a constant.  You can pass linesearch={float,np.ndarray} to optimize functions and it will use this in the optimizer iterations.

msricher commented on flik/optimize/functions.py:
```diff
@@ -0,0 +1,512 @@
+def _root_iteration(f, x, J, tol, maxiter, method, linesearch, trustregion):
+    r"""Run the iterations for ``root``."""
+    # Set up working data for direct methods
+    method.setup()
```
Multi-secant and direct methods (i.e. not using any Jacobian) require the allocation of some internal, working data.  Method.setup() and Method.cleanup() at the end of this function ensure that this stuff is handled by anyone making a Method subclass.

msricher commented on flik/optimize/functions.py:
```diff
@@ -0,0 +1,512 @@
+        Jacobian of function ``f``, ``J(x)`` if callable or Jacobian instance.
+        Finite difference method ("forward" | "central") if str.
+        By default, a central difference Jacobian will be used where necessary.
+    cost : callable, optional
```
For example, `cost` might be the energy function, if you have some nonlinear system of equations corresponding to a (variational) wavefunction optimization.

msricher commented on flik/methods/diis.py:
```diff
@@ -0,0 +1,59 @@
+    ]
+
+
+class DiisMethod(Method):
```
Can we port Toon's code from HORTON 2?

matt-chan replied:
The romin project has cleaner DIIS code that we can use almost verbatim.

gabrielasd committed:
Add Pearson and McCormick updates

msricher committed:
Give linesearch & trustregion access to Hessian matrix in API

msricher commented on flik/trustregion/base.py:
```diff
@@ -50,7 +50,7 @@ def __init__(self, f, J):
         self._jacobian = J
 
-    def __call__(self, dx, x, b):
+    def __call__(self, dx, x, b, A):
```
API is compatible with laurent/cristina's code

msricher committed:
fix dfpinv method

Commit:
revert back to version.py usage

Commit:
Add Hoshino update

msricher committed:
Cleanup DFP/SR1INV and BOFILL updates

msricher committed:
Add test generators for Quasi-Newton methods

Update tests for structure

Add more tests

Commit:
Add ss-bfgs

msricher committed:
Add tests for new quasi-newtons (some FAILING)

xiaominhc committed:
Added Xie's and Yuan's jac and invjac updates.

Added tests for Xie and Yuan (FAILING ALL TESTS)

msricher committed:
Change setup.cfg to force serial processing and output with `nosetests -s`

Revert test stuff that was fine before I realized this

msricher committed:
Fix style, use numpy.testing error messages to see which methods fail

msricher committed:
Add benchmark function package

msricher committed:
missing numpy import

kimt33 commented:
So I went through everything and I made some notes on what I think. I think we'll discuss it in person on Monday, but for transparency, I'll just write down what I have here.

I'll start with the structures that I agree with:
1.`TrustRegion` and `LineSearch` modules are separate modules, independent of the other modules. 
2. `root` and `minimize`  act as the UI for most users (analogous to the ones in scipy)
3. separate structure for different quasinewton and newton algorithms

Following are some of the concerns (in decreasing importance)
1. I think there should be a class that contains information for each point (e.g. objective value, gradient, hessian, etc) at different iterations. I think this is conceptually nicer (the things we are optimizing should be separated from how we are optimizing) and I think it will be cleaner when we need to implement algorithms that depend on multiple iterations (e.g. diis, multisecant) rather than hardcoding them somewhere.
2. `Method` class is a weird mix of class structure and procedural structure. There are a lot of static methods that seem to simply be called once. I get that different quasinewton algorithms use similar methods (like the step methods) and maybe a "library" of step update methods might be nice to have, but all of the update methods seem to be called only once. These should be moved to its own class that overwrites the update method
3. `_root_iteration` hardcodes a specific way of using the given `Method`, `TrustRegion`, and `LineSearch` instances. There is not much room for flexibility without going to this method and changing things. I like that there is an interface between the modular components and the overarching algorithm that brings it all together, but the current implementation is not flexible or transparent enough to make this easy (I think). 
4. I'm not sure if any of the private attribute/method is necessary. They really do nothing (in python) and if we are going to be making a porcelain part anyways, a normal user will have no business looking at the plumbing. I thought that the way that `Method.update` simply calls a private method is really odd.
5. I'm not too sure about creating a multiple-inheritance structure for handling the creating gradient and jacobians with different finite difference algorithms. As things are implemented right now, it seems that they do not even need to be classes at all, since these are only used in the initialization of `root` and `minimize`. A wrapping function could have been used. I'm not sure how much the class structure will help with the implementation. I mean, I don't mind the class structure (just from conceptual aspect), I thought that the multiple inheritances seem unnecessary. If we have a point class, we can move it there.
6. I don't think the `__call__` is necessary for `TrustRegion` and `LineSearch`. It makes sense for the `Gradient` and the `Jacobian` because we evaluate these, but `TrustRegion` and `LineSearch` should probably have a separate method dedicated to solving them for clarity.
7. I think the initialization aspect should either be handled with class structure or with straight up functions rather than a function that calls functions defined outside of the class. There seem to be some  hidden behaviors that you can't find without going through all of the code. If it is a single function (w/o calling other functions) then it will be longer, but it will be easier to read. The code repetition doesn't seem to be that big of a problem if there's only two functions. But yeah, this, to me, has the lowest priority since porcelain should probably be built after the plumbing is stable.

@tczorro seems to have a structure in his optimizer that (I think) resembles your structure except for some of the things I commented on. I will to look at his code before Monday so I have a better idea about it. But yeah, I think your structure is alright (it covers most of what I would like to have) but I think we need to properly discuss what we want out of the code so that we can make design decisions accordingly.

msricher commented:
Response to David's comments:
1) A point class would be fine.  I didn't bother, but I don't care either way.  I figured the DIIS method or other multi-secant methods that keep history would be responsible for storing the data it needs across iterations.
2) I didn't want to write 20+ subclasses for the basic quasi-Newton methods.  I like the structure (other than this part), so I'm not sure what to do about it.  Maybe make 20+ subclasses.  Maybe leave it be.
3) More routines can be written (such as saddle point optimization)
4) Again the alternative is 20+ subclasses.
5) I used this to make sure I always have a callable Jacobian, finite difference if necessary.  This structure will probably never have to be changed again. A simple function wrapper would also work just as well.  /However/, as it is now, it can be used for checking that certain methods will only work with Gradient/Hessian, and that certain methods will only work with Objective/Jacobian.
6) `__call__` is just a name, it can be changed.
7) Don't care, can be restructured as David said.

matt-chan commented:
To add to the discussion:
1. I also thought the point class would be within DIIS/multi-secant classes. Maybe we can make a shared one (it's technically just a list or two...), but it adds extra fixed API elements. 
2. I think 20 subclasses would be clearer, and less bug prone. We should only ever hit one of the `elif`s and it shouldn't matter what the order of evaluation is. As it stands, if we have a bad (or redundant) condition in an earlier `elif`, we won't ever hit the lower ones. This can be something we fix in the "master's sprint".
3. @kimt33 What are the other types of problems we'll have to solve with `_root_iteration`? If it's only a few more, we could just write parallel methods and select based on the type of problem from `root`
4. I'd prefer the subclasses here as well. It's hard to see at first glance which methods are used where. I'm not as committed to this as the `Method` class though. 
5. Can we talk about this one in person? I don't completely understand why we're using multiple inheritance here. 
6. Let's change it in the sprint.
7. @kimt33 I don't understand which part of the code you're referring to. Can you give an example? 

And I'll add two more points.... 
8. I think `lsopts` and `tropts` should be removed from `root` and `minimize` and the options provided to the line search and trust region when initialized instead. It'll break introspection and documentation lookup when we provide them separately from instantiation, and also it adds some complexity to the methods. The only benefit from having them is that we get to provide additional options to line search/trust region methods by string instead of by code, which is not worth it I think.
9. (minor) We should make sure the Integral type checks print out warnings asking for integers (etc). I think most users won't know what an "Integral" type is.

kimt33 commented:
1. I think it is necessary to have it as a separate class. I think it helps a lot with the organization and moving it to the DIIS seems to be a hardcoded solution to avoid implementing a point class.
2. I don't mind the `step` methods being static, but I think that moving `update` methods will make things a bit clearer. Not a big change.
3. Ask derrick. But from a conceptual side, there are many ways we can group together different optimization routines into a single "optimizer". For example, I would like to optimize my orbitals with a different optimizer than the optimizer for the wavefunction. For my case, it is fairly simple because I just run one after the other, but I can easily imagine it becoming more complicated and I would either need to hardcode these cases or come up with some hack to go around the limitations of the optimizer.
7. I am talking about the `flik.optimize.functions`

matt-chan commented:
7. I see. Yes, there is a little bit of non-modularity there. Let's fix that in the design meeting if we have time. It's porcelain, but it still touches quite a few unit tests etc. I'm not sure how many other "master sprints" we'll have before going production.

Let's talk about 3. tomorrow?

Commit:
Add Example test class

Commit:
Add import statements to tests

Commit:
Delete example 1

Commit:
Fix error

Commit:
Sec posdef test (#34)

* Add positive definite test

incomplete hoshino_inv doesn't work

* Generalize posdef test in any dimension

* Add secant condition

Incomplete

* Add positive definite test

incomplete hoshino_inv doesn't work

* Generalize posdef test in any dimension

* Add secant condition

Incomplete

* Change f1 to Example1 class

xiaominhc committed:
More quasiupds (#35)

* Check that grad/hess used for minimize

* fix dfpinv method

* missing numpy import

* Added Xie's and Yuan's jac and invjac updates.

* Added tests for Xie and Yuan (FAILING ALL TESTS)

* Fix DFP inv (also BFGS) update. Fix Xie/Xie inv.

* Fix Hoshino (and inv) update.

* Fixed Xie(inv) and Yuan(inv) (pass test_base).

gabrielasd committed:
Pearson mccormick (#36)

* Add Pearson and McCormick updates

* Add Pearson and McCormick updates

* Added Xie's and Yuan's jac and invjac updates.

* Added tests for Xie and Yuan (FAILING ALL TESTS)

* Fix code inconsistency for Parson and McCormick updates

* Fix DFP inv (also BFGS) update. Fix Xie/Xie inv.

* Fix McCormick inverse update

* Clean McCormick inverse update

gabrielasd committed:
Fix dfp bfgsinv (#37)

* Check that grad/hess used for minimize

* Fix DFP and BFGS inverse updates

* Test Method.update for this methos passes
